### ISSUE TITLE
[HOLD] [New Feature] Implement URL Previews

### ISSUE THREAD

**Issue by marcaaron** [Expensify/App#15001]
Coming from https://github.com/Expensify/Expensify/issues/149670

cc @wojtus7

### Implement Link Previews in New Expensify

**Problem:**

When a user shares a traditional “blue link” URL with chat participants in New Expensify they require an additional interaction from the participants to figure out whether the content is relevant to them. This is a drag on productivity and also forces the user to leave the app or switch context. When presented with links like these users might not feel a strong urge to interact with them at all.

**Solution:**

A common solution for this is to include some form of an expandable “link preview” as a convenience (no need to leave the chat) while providing quick context and increasing the likelihood that people will interact with the content.

**Slack thread:**

https://expensify.slack.com/archives/C01GTK53T8Q/p1676052871304119

Slack: 
![image (1)](https://user-images.githubusercontent.com/1371996/103492781-e8e9e880-4df2-11eb-8b6b-178229478070.png)

iMessage: 
![image (10)](https://user-images.githubusercontent.com/1371996/103492789-f4d5aa80-4df2-11eb-85cf-3d0ea89f90ad.png)

WhatsApp: 
![image (11)](https://user-images.githubusercontent.com/1371996/103492787-f2735080-4df2-11eb-98e7-225a4e34813d.png)

## Tasks

- [x] Post **Proposal (full Problem/Solution statement)** in `#expensify-open-source`
- [x] Wait at least one full business day, and until the post has a majority (2/3) of positive reactions (👍)
- [x] Paste Proposal in the space above with a link to the Slack thread
- [ ] Email `strategy@expensify.com` and paste in the Proposal
- [x] Fill out the **High-level overview of the problem**, **Timeline**, and **Terminology** sections of the Design Doc
- [ ] Email `strategy@expensify.com` (continue the same email chain as before) with the link to your Design Doc
- [x] Host a **pre-design** meeting ([example](https://expensify.slack.com/archives/C01GTK53T8Q/p1665794669539419)) in `#expensify-open-source` to discuss any necessary details in public before filling out the **High-level of proposed solution** section.
- [x] Fill out the **High-level of proposed solution** section
- [ ] Email `stategy@expensify.com` again with links to the doc and pre-design conversation in Slack
- [ ] Add the `DesignDocReview` label to get the **High-level of proposed solution** section reviewed
- [ ] Respond to any questions or concerns and bring up blockers in Slack to get a consensus if necessary
- [ ] Confirm that the doc has the minimum necessary number of reviews before proceeding
- [ ] Host another pre-design meeting in `#expensify-open-source` to ask for engineering feedback on the technical solution.
- [ ] Fill out the **Detailed implementation of the solution** and related sections.
- [ ] Re-add the `DesignDocReview` label to this issue
- [ ] Respond to any questions or concerns and bring up blockers in Slack to get consensus if necessary
- [ ] Confirm that the doc has the minimum necessary number of reviews before proceeding
- [ ] Email `strategy@expensify.com` one last time to let them know the Design Doc is moving into the implementation phase
- [ ] Implement the changes
- [ ] Send out a follow up email to `strategy@expensify.com` once everything has been implemented and do a **Project Wrap-Up** retrospective that provides:
    - Summary of what we accomplished with this project
    - What went well?
    - What could we have done better?
    - What did we learn?

**Comment by marcaaron** [comment]
We've already got [some of the scaffolding in the API endpoint set up](https://github.com/Expensify/Web-Expensify/blob/c697a739c0ab666a2c84ea2be7a1b63b23430eaa/api.php#L702-L704) to fetch the metadata itself, but I think the original MVP design was pretty underwhelming so gonna think about a new solution...

Basic high level design...

- I think we'd want to support multiple links...
- The metadata should be returned in the `reportAction.message`, but not necessarily stored on it.
- We can resolve it and return it when syncing `reportActions` or pushing them to clients.
- We should get the metadata itself from a cache or create and cache url metadata when a comment is created that has a link in it.
- We can then cache the url metadata with a key of url? If someone shares the same URL we won't have to get the metadata again. `[url]: {metadata: {...}, lastFetched: <timestamp>}` (maybe just use nvps for this?)
- If a new comment is created with the same url we will check to see when it was `lastFetched` and maybe re-fetch it or else just return the cached version. And we can have some kind of reasonable default length of time (1 week) to trigger the re-fetch.

**Later improvements**

- Clients could technically end up with previews that are out of date and cached locally at some point. This doesn't seem too important to fix yet. But we could future proof this by including the `lastFetched` in every link meta data preview. If a client tries to render the preview then we trigger it an update in the server then push out the new metadata (if it has changed) or just update the `lastFetched`.

**Comment by marcaaron** [comment]
Started a predesign here: https://expensify.slack.com/archives/C01GTK53T8Q/p1676659234489209

Looks like we are going to work with SWM on this one and most likely use a 3rd party API so our internal team can prioritize road map features over this "nice to have" feature.

**Comment by marcaaron** [comment]
Going to do this one a bit looser and jump straight into the detailed design based on the predesign. 

**Why?** The conversations were relatively uncontroversial, the feature is relatively simple, and alternatives (in the above comments) were dismissed during the predesign when we settled on a 3rd party API path. Additionally, people are quite busy at the moment with more important docs and ECX. So, to make this "worth it" I'm following the process slightly more loosely than usual and working with Software Mansion on the implementation.

**What does it mean exactly?** We're still going to use the Design Doc template to plan this project. But possibly will jump to the implementation if there's no major pushback on the general plan.

**Comment by marcaaron** [comment]
Currently looking into the https://microlink.io API service. 
Going to start by setting up a free endpoint for @wojtus7 to test with.

**Comment by marcaaron** [comment]
Endpoint is on production now. Going to reach out via Slack with more details.

**Comment by marcaaron** [comment]
@wojtus7 from SWM left a status update [here](https://expensify.slack.com/archives/C04878MDF34/p1679412429196809?thread_ts=1676065837.626589&cid=C04878MDF34) and is working on this. It's looking pretty good.

**Comment by MelvinBot** [comment]
Triggered auto assignment to @strepanier03 (`NewFeature`), see https://stackoverflowteams.com/c/expensify/questions/14418#:~:text=BugZero%20process%20steps%20for%20feature%20requests for more details.

**Comment by strepanier03** [comment]
Will start working on this tomorrow, missed it in my k2.

**Comment by marcaaron** [comment]
Oh actually @strepanier03 we are well underway with the feature. I added the label for posterity, but we are good on the triage front! 🙇

**Comment by marcaaron** [comment]
This is still in development

**Comment by marcaaron** [comment]
We've got a PR up now where we are exploring this addition

https://github.com/Expensify/App/pull/17288

**Comment by thienlnam** [comment]
I have read and reviewed this Design Doc!

**Comment by marcaaron** [comment]
I think the latest update is here: https://expensify.slack.com/archives/C04878MDF34/p1683241460982269

I saw the PR was recently updated by @wojtus7 so perhaps we will make some progress there soon?

**Comment by marcaaron** [comment]
Just a heads up, I'm going to be OOO from May 23rd to May 30th. I noticed that @AndrewGable is also assigned to the review though so feel free to proceed without me if you have time in your schedule - otherwise, I'll review when I get back!

**Comment by JmillsExpensify** [comment]
Quickly chiming in on that last comment, I think this one is fine to let this one go while you're on vacay. This is primarily a polish item within the app.

**Comment by strepanier03** [comment]
@sakluger - I am heading out on sabbatical and this GH won't be resolved and ready for any BZ action until after I'm already gone. 

Thank you for taking it over. For now it's just monitoring and waiting for this to be ready for a BZ to take any action.

**Comment by wojtus7** [comment]
Hi

Since Link Preview was merged I think much needed follow up is feature to hide previews. But hiding them only in the component itself seems strange since user would need to hide unwanted preview every single time when entering the chat. 
@marcaaron do you think we can create some API call with `reportActionID` and links that shouldn't be returned in preview info response? That way we can hide it for everyone in conversation.
But that also create problem of who can hide previews etc.

Or should I hide them for now locally using something like async storage?

**Comment by sakluger** [comment]
@strepanier03 @marcaaron just confirming that I don't need to take any action here, right? Do you anticipate me needing to take any action at any point?

**Comment by marcaaron** [comment]
@sakluger we're good for now yeah thanks!

@wojtus7 great question. I think we'd add an API call like you mentioned and allow the link preview to be "hidden". And once hidden I think we could would stop calling the API. Let me ask the team and look into what setting that up would take and get back to you. Sounds like a good next issue for us to collaborate on.

**Comment by marcaaron** [comment]
Gonna drop this to a weekly. I won't be able to prioritize it at a Daily frequency until sometime next week.

**Comment by iwiznia** [comment]
FYI this PR https://github.com/Expensify/App/pull/17288 had this bug https://github.com/Expensify/App/issues/21368 (I already sent a PR for it)

**Comment by sakluger** [comment]
Thanks for assigning @iwiznia

**Comment by marcaaron** [comment]
Adding the API keys here -> https://github.com/Expensify/Web-Expensify/pull/38045/files

This should fix the rate limiting. And then we can move on to the CSP issues - which are unexpected and I am not entirely sure how to solve yet.

**Comment by marcaaron** [comment]
Link previews are live now 🎉  But we still need to:

- [ ] Figure out the CSP issues on web
- [ ] Implement a new command to hide the link preview
- [ ] Hook up the front end for hiding the link preview

**Comment by sakluger** [comment]
@marcaaron what's the latest on this one?

**Comment by marcaaron** [comment]
Don't have much to report other than...

- We have slowly been bumping up the rate limit for the Microlink API to handle the current capacity
- Web is still broken because of CSP issues I haven't had time to look into
- We are discussing implementing a "hide" feature
- Overall this is a lower priority project that is blocked on a couple of Internal issues that I'll get to whenever I get to them.

**Comment by marcaaron** [comment]
This is probably more of a monthly for me than a weekly at this point.

**Comment by marcaaron** [comment]
This is kind of on perpetual HOLD at the moment. The feature was not useful enough and annoyed people. So, we disabled it temporarily to prioritize some more valuable features. We will probably return to improve it at some point in the future. But not yet.

**Comment by marcaaron** [comment]
Cool. I see the issue also mentions that it is disabled. So, we can fix it if/when we re-enable it. I maybe wouldn't spend much time on this.

**Comment by sakluger** [comment]
In order to roll back out, I think we'll need some more design discussions and rework of the feature. With our current focus on N5-9, should we close the issue and revisit at a later date, or should we leave the issue open and on hold indefinitely?

**Comment by marcaaron** [comment]
Yes, let's be honest. This was a bit of a failure to pre-design some things. But we have some code in hand - which ain't the worst situation to be in 😄 

Could end up in a vip, but unlikely a wave. I am fine with holding it until we decide it doesn't belong in any waves or VIPs. 

That said, we know we will need it. It needs a design doc so we can properly plan hiding link previews and other edge cases (I am happy to help write that eventually or help someone else write it). It feels like a "must have" for vsb chat adoption to me.

**Comment by marcaaron** [comment]
Actually let's just close this. It's distracting. We can reopen this conversation later. I'm going to keep the Project with all the issues open for now.
